### ISSUE TITLE
scmp: does traceroute work as expected?

### ISSUE THREAD

**Issue by lryll** [scionproto/scion#1912]
When running `scmp` with `traceroute` commandI get the following output
```
$ ./bin/scmp tr -local 19-ffaa:1:b5,[10.0.8.32] -remote 18-ffaa:1:123,[10.0.8.19]
INFO[10-01|10:56:02] Started                                  id=869153a8 goroutine=dispatcher_bck
Using path:
  Hops: [19-ffaa:1:b5 1>46 19-ffaa:0:1303 1>5 19-ffaa:0:1301 3>5 18-ffaa:0:1201 6>1 18-ffaa:0:1202 34>1 18-ffaa:1:123] Mtu: 1472
0  19-ffaa:1:b5,[127.0.0.1] IfID=1  655µs 150µs 146µs
1  * * *
2  * * *
3  * * *
4  * * *
5  * * *
6  * * *
7  * * *
8  * * *
9  * * *
10  * * *
```
Is this the expected output? I thought it would print more information about the hops.
(Note: I am connected to SCIONLab with openvpn)
Also, it prints `19-ffaa:1:b5,[127.0.0.1]`. Shouldn't this be  `19-ffaa:1:b5,[10.0.8.32]`?

**Comment by sgmonroy** [comment]
Could you paste your local topology file? It might explain why you are seeing the first hop address as 127.0.0.1.

Could you also run scmp echo and rp (recordpath) to some remote address in 19-ffaa:0:1303 (should be your parent AS) ?

**Comment by sgmonroy** [comment]
Yes, with traceroute the local/internal address of the border router is the one being shown.

When you say you get the same output with echo and recordpath, do you mean no reply?

I am not very familiar with the setup you are running, could you paste your interface addresses?

You also mentioned that you control the AS 19-ffaa:1:122, could you also do the ping/tr/rp from that AS to AS 19-ffaa:1:b5 ?

**Comment by lryll** [comment]
> When you say you get the same output with echo and recordpath, do you mean no reply?

Yes, exactly.

>I am not very familiar with the setup you are running, could you paste your interface addresses?

My setup is two Ubuntu 16.04 VMs (the first below installed via vagrant the other manually from ISO).
`19-ffaa:1:122,[10.0.8.22]` - [network interfaces](https://gist.github.com/lryll/454e96c9b293108299d7da381966ec63)
`19-ffaa:1:b5,[10.0.8.32]` - [network interfaces](https://gist.github.com/lryll/faa96d66c95152f67ad45788f62a9693)

I supposed with interface addresses you meant the network interfaces?
**traceroute:**
```
./bin/scmp tr -local 19-ffaa:1:122,[10.0.8.22] -remote 19-ffaa:1:b5,[10.0.8.32]
INFO[10-03|08:59:51] Started                                  id=8d2f5bf6 goroutine=dispatcher_bck
Using path:
  Hops: [19-ffaa:1:122 1>36 19-ffaa:0:1303 46>1 19-ffaa:1:b5] Mtu: 1472
0  19-ffaa:1:122,[10.0.2.15] IfID=1  642µs 408µs 233µs
1  * * *
2  * * *
3  * * *
4  * * *
```
**recordpath:** (interesting ERROR though)
```
./bin/scmp rp -local 19-ffaa:1:122,[10.0.8.22] -remote 19-ffaa:1:b5,[10.0.8.32]
INFO[10-03|09:00:46] Started                                  id=45c2a733 goroutine=dispatcher_bck
Using path:
  Hops: [19-ffaa:1:122 1>36 19-ffaa:0:1303 46>1 19-ffaa:1:b5] Mtu: 1472
ERROR: read unix @->/run/shm/dispatcher/default.sock: i/o timeout

```
**echo** (returns after >1min)
```
./bin/scmp echo -local 19-ffaa:1:122,[10.0.8.22] -remote 19-ffaa:1:b5,[10.0.8.32]
INFO[10-03|09:01:38] Started                                  id=74b59c8b goroutine=dispatcher_bck
Using path:
  Hops: [19-ffaa:1:122 1>36 19-ffaa:0:1303 1>5 19-ffaa:0:1301 5>1 19-ffaa:0:1303 46>1 19-ffaa:1:b5] Mtu: 1472
--- 19-ffaa:1:b5,[10.0.8.32] statistics ---
90 packets transmitted, 0 received, 100% packet loss, time 1m29.29073s
```

The behavior is the same for when I run the commands in the other VM. Though, independent from  what's the local AS and what's the remote AS.

So it looks like the problem is connected to the socket error of the dispatcher (`ERROR: read unix @->/run/shm/dispatcher/default.sock`. )?

**Comment by juagargi** [comment]
So far I have not seen `echo` or `traceroute` not working. 
Running from a dedicated machine in `17-ffaa:1:a` in SCIONLab:
```
$ ./bin/scmp tr -local 17-ffaa:1:a,[127.0.0.1] -remote 19-ffaa:0:1303,[127.0.0.1]
INFO[10-03|13:45:02] Started                                  id=4311ace1 goroutine=dispatcher_bck
Using path:
  Hops: [17-ffaa:1:a 1>147 17-ffaa:0:1107 1>4 17-ffaa:0:1102 3>3 17-ffaa:0:1103 4>8 17-ffaa:0:1101 5>4 18-ffaa:0:1201 5>3 19-ffaa:0:1301 5>1 19-ffaa:0:1303] Mtu: 1472
0  17-ffaa:1:a,[127.0.0.1] IfID=1  288µs 187µs 101µs
1  17-ffaa:0:1107,[10.0.8.1] IfID=147  1.22ms 1.204ms 1.292ms
2  * * *
3  17-ffaa:0:1102,[192.33.93.190] IfID=4  3.237ms 2.902ms 2.91ms
4  * * *
5  17-ffaa:0:1103,[192.33.92.69] IfID=3  4.998ms 4.585ms 4.528ms
6  * * *
7  17-ffaa:0:1101,[193.247.172.154] IfID=8  9.912ms 9.461ms 9.741ms
8  * * *
9  18-ffaa:0:1201,[128.237.152.180] IfID=4  119.997ms 118.541ms 119.184ms
10  * * *
11  19-ffaa:0:1301,[141.85.227.185] IfID=3  257.793ms 256.543ms 257.003ms
12  * * *
13  19-ffaa:0:1303,[141.44.25.144] IfID=1  293.991ms 294.497ms 293.972ms
14  19-ffaa:0:1303,[127.0.0.1]  294.895ms 293.652ms 292.775ms
```
Echo also works:
```
$ ./bin/scmp echo -local 17-ffaa:1:a,[127.0.0.1] -remote 19-ffaa:0:1303,[127.0.0.1]
INFO[10-03|13:48:39] Started                                  id=7eadc7d8 goroutine=dispatcher_bck
Using path:
  Hops: [17-ffaa:1:a 1>147 17-ffaa:0:1107 1>4 17-ffaa:0:1102 3>3 17-ffaa:0:1103 4>8 17-ffaa:0:1101 5>4 18-ffaa:0:1201 5>3 19-ffaa:0:1301 5>1 19-ffaa:0:1303] Mtu: 1472
176 bytes from 19-ffaa:0:1303,[127.0.0.1] scmp_seq=0 time=293.15ms
176 bytes from 19-ffaa:0:1303,[127.0.0.1] scmp_seq=1 time=293.565ms
^C
--- 19-ffaa:0:1303,[127.0.0.1] statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1.592633s
```
I am going to see if there is any difference when running `scmp` inside a VM, for whatever reason, and I'll report back.

**Comment by juagargi** [comment]
Now from a VM, in `19-ffaa:1:d` also in SCIONLab:
```
$ ./bin/scmp echo -local 19-ffaa:1:d,[127.0.0.1] -remote 17-ffaa:1:a,[127.0.0.1]
INFO[10-03|12:24:38] Started                                  id=ca6f7af3 goroutine=dispatcher_bck
Using path:
  Hops: [19-ffaa:1:d 1>34 19-ffaa:0:1303 1>5 19-ffaa:0:1301 7>1 19-ffaa:0:1302 2>11 17-ffaa:0:1101 8>4 17-ffaa:0:1103 3>3 17-ffaa:0:1102 4>1 17-ffaa:0:1107 147>1 17-ffaa:1:a] Mtu: 1472
184 bytes from 17-ffaa:1:a,[127.0.0.1] scmp_seq=0 time=139.875ms
184 bytes from 17-ffaa:1:a,[127.0.0.1] scmp_seq=1 time=139.453ms
184 bytes from 17-ffaa:1:a,[127.0.0.1] scmp_seq=2 time=139.158ms
^C
--- 17-ffaa:1:a,[127.0.0.1] statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2.937329s
```
Traceroute works as well. Let's focus on `echo`.
I see the main difference is the IP used. I am using the `lo` one, @lryll is using the one from `tun0`. When running `echo` with my VPN IP, or the other side VPN IP, I also see no responses:
```
$ ./bin/scmp echo -local 17-ffaa:1:a,[10.0.8.116] -remote 17-ffaa:0:1107,[127.0.0.1] -interval 10ms
INFO[10-03|14:33:02] Started                                  id=3d7067f7 goroutine=dispatcher_bck
Using path:
  Hops: [17-ffaa:1:a 1>147 17-ffaa:0:1107 1>4 17-ffaa:0:1102 3>3 17-ffaa:0:1103 4>8 17-ffaa:0:1101 8>4 17-ffaa:0:1103 2>2 17-ffaa:0:1102 4>1 17-ffaa:0:1107] Mtu: 1472
^C
--- 17-ffaa:0:1107,[127.0.0.1] statistics ---
751 packets transmitted, 0 received, 100% packet loss, time 7.508873s
```
But `tcpdump` running on the only machine at `17-ffaa:0:1107` confirms the packets arrive to the other side:
```
$ sudo tcpdump -n -i tun0 src 10.0.8.116
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on tun0, link-type RAW (Raw IP), capture size 262144 bytes
14:33:04.608016 IP 10.0.8.116.50000 > 10.0.8.1.50146: UDP, length 160
14:33:04.617900 IP 10.0.8.116.50000 > 10.0.8.1.50146: UDP, length 160
14:33:04.627947 IP 10.0.8.116.50000 > 10.0.8.1.50146: UDP, length 160
14:33:04.637925 IP 10.0.8.116.50000 > 10.0.8.1.50146: UDP, length 160
14:33:04.647893 IP 10.0.8.116.50000 > 10.0.8.1.50146: UDP, length 160
14:33:04.658371 IP 10.0.8.116.50000 > 10.0.8.1.50146: UDP, length 160
14:33:04.667777 IP 10.0.8.116.50000 > 10.0.8.1.50146: UDP, length 160
14:33:04.678132 IP 10.0.8.116.50000 > 10.0.8.1.50146: UDP, length 160
14:33:04.688432 IP 10.0.8.116.50000 > 10.0.8.1.50146: UDP, length 160
14:33:04.698204 IP 10.0.8.116.50000 > 10.0.8.1.50146: UDP, length 160
14:33:04.708087 IP 10.0.8.116.50000 > 10.0.8.1.50146: UDP, length 160
14:33:04.730715 IP 10.0.8.116.50000 > 10.0.8.1.50146: UDP, length 160
14:33:04.730729 IP 10.0.8.116.50000 > 10.0.8.1.50146: UDP, length 160
14:33:04.738023 IP 10.0.8.116.50000 > 10.0.8.1.50146: UDP, length 160
14:33:04.748019 IP 10.0.8.116.50000 > 10.0.8.1.50146: UDP, length 160
```

**Comment by lryll** [comment]
Yes, I could get the program runnnig and everything works fine when using `127.0.0.1` as ip.

What I still don't understand is: Why the localhost ip and not the vpn ip address is used (or in case of no vpn usage the real ip address) in **scmp**.
(So I experienced the same as in the [last post](https://github.com/scionproto/scion/issues/1912#issuecomment-426621365) of @juagargi above.)

**Comment by matzf** [comment]
@lryll, @sgmonroy : I've just had another look at this as well. What I've observed is that the border router logs 
```
2018-11-05 12:57:01.626234+0000 [EROR] Error processing packet rpkt=b063915b err="Unable to process unsupported payload type pldType="<nil>" pld="<nil>""
```
for every echo request sent. Digging into the source code (in the `netsec-ethz/netsec-scion` `scionlab`-branch), I found this:
https://github.com/netsec-ethz/netsec-scion/blob/0336818c24907ac60b73ba45bf4f69b528a870ed/go/border/rpkt/process.go#L100-L103

As the FIXME comment indicates, the border router seems to hijack any SCMP packet sent to its IP, regardless of the actual destination port. By setting the `-local` address to the VPN IP, this behaviour is triggered and the packet before being returned to `scmp`.

This issue appears to have been addressed in #1797.

**Comment by sgmonroy** [comment]
AFAIK BR never replied to scmp echo, it does reply to scmp traceroute, and add info in the packet for scmp recordpath.

Currently the dispatcher is the one replying to scmp echo/traceroute/recordpath, so that means that the target ip has to be a service address or anything running on top of a dispatcher.
The local address should also be in address similar to any other service, at the end of the day, the scmp tool behaves like a service and runs on top of the dispatcher.
So your local address should be an address on the host you are running the scmp tool from, not the border router external interface address.

As @matzf has mentioned, #1797 reworked the border router. Before the change, a packet with destination address as one of the interface addresses would be tagged as destined to the border router, and as @matzf showed in the source code link, there was no support to deal with it.
So probably an easy fix would be to just forward the packet (ie. remove the *scmp.Hdr case), so it would reach the dispatcher.

After #1797, it is still not possible to ping a border router except the Control address, which basically is a service running on top of the dispatcher.
